lib/lrmd/lrmd_output.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2020 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include <stdarg.h>
+
+#include <crm/lrmd_internal.h>
+#include <crm/common/output_internal.h>
+
+static int
+default_list(pcmk__output_t *out, lrmd_list_t *list, int rc, const char *title) {
+    lrmd_list_t *iter = NULL;
+
+    if (rc <= 0) {
+        return pcmk_rc_error;
+    }
+
+    out->begin_list(out, NULL, NULL, "%s", title);
+
+    for (iter = list; iter != NULL; iter = iter->next) {
+        out->list_item(out, NULL, "%s", iter->val);
+    }
+
+    out->end_list(out);
+    lrmd_list_freeall(list);
+    return pcmk_rc_ok;
+}
+
+static int
+xml_list(pcmk__output_t *out, lrmd_list_t *list, int rc, const char *ele) {
+    lrmd_list_t *iter = NULL;
+
+    if (rc <= 0) {
+        return pcmk_rc_error;
+    }
+
+    for (iter = list; iter != NULL; iter = iter->next) {
+        pcmk__output_create_xml_text_node(out, ele, iter->val);
+    }
+
+    lrmd_list_freeall(list);
+    return pcmk_rc_ok;
+}
+
+PCMK__OUTPUT_ARGS("alternatives-list", "lrmd_t *", "const char *")
+static int
+lrmd__alternatives_list_xml(pcmk__output_t *out, va_list args) {
+    lrmd_t *lrmd_conn = va_arg(args, lrmd_t *);
+    const char *agent_spec = va_arg(args, const char *);
+
+    lrmd_list_t *list = NULL;
+    int rc;
+
+    xmlNodePtr node = pcmk__output_xml_create_parent(out, "providers");
+
+    xmlSetProp(node, (pcmkXmlStr) "for", (pcmkXmlStr) agent_spec);
+
+    rc = lrmd_conn->cmds->list_ocf_providers(lrmd_conn, agent_spec, &list);
+    return xml_list(out, list, rc, "provider");
```

Comment:
I agree on not having side effects in the output messages.  This seems like another thing to audit at some point.

I could do a single set of custom messages, but that would get rid of some of the extras I added.  The alternatives message adds an attribute for the agent it's listing alternatives for, for instance.  The agents message adds standard and provider attributes.  For crm_resource, you could say it doesn't matter because you'd already know these based on the command line arguments.  However, they seemed useful enough to have in an external library where you may want that kind of information.  I can do whatever, though.

Reply:
Separate custom messages would also be fine, or a single message could take a whole bunch of optional arguments.